### ISSUE TITLE
ttnn.xlogy_bw and ttnn.addcdiv_bw hard-code the sign of the infinity at a zero divisor and skip the NaN case

### ISSUE THREAD

**Issue by ping-dev-ui** [tenstorrent/tt-metal#55316]
**Component / Area:** ttnn eltwise binary/ternary backward (composite), `xlogy_bw`, `addcdiv_bw`

**Issue Type:** Bad Outputs

## Summary

Both ops special-case a zero divisor with a constant-sign infinity instead of the sign of the actual quotient, and `xlogy_bw` omits the `x == 0` NaN arm entirely.

`xlogy_bw`, gradient with respect to `other`
([binary_backward.cpp#L295-L306](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward.cpp#L295-L306)):

```cpp
grad2 = grad * input_a * reciprocal(other);
grad2 = where(eqz(other), sign(grad) * inf, grad2);   // ignores sign(input_a); no NaN arm for input_a == 0
```

Torch's derivative is `grad * self / other`, so at `other == 0` the sign is `sign(grad * self)` and the value is NaN when `self == 0`.

`addcdiv_bw`, gradients with respect to `tensor1` and `tensor2`
([ternary_backward.cpp#L52-L76](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/cpp/ttnn/operations/eltwise/ternary_backward/ternary_backward.cpp#L52-L76)):

```cpp
grad_a = grad * value * reciprocal(tensor2);
where(eqz(tensor2), where(eqz(grad), nan, +inf), grad_a)      // ignores sign(grad) and sign(value)
grad_b = (-grad * value * tensor1) * reciprocal(square(tensor2));
where(eqz(tensor2), where(eqz(grad), nan, -inf), grad_b)      // ignores sign(grad * value * tensor1); no NaN for tensor1 == 0
```

Torch: `tensor1: grad * value / tensor2` and `tensor2: -grad * value * tensor1 / tensor2^2`, signed by the full product, NaN when the numerator is zero.

`div_bw` in the same file already does this correctly
([binary_backward.cpp#L723-L740](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward.cpp#L723-L740)):
`sign(input_a) * neg_inf * sign(grad)` with a NaN arm for `input_a == 0`. These two ops predate that fix and were never aligned.

This is a source-level analysis with a host replay; I do not have a card. The `where`/`sign` chain is exact, so the device result at a zero divisor is the constant the code selects.

## Observed vs expected

`xlogy_bw`, `y == 0`:

| x | grad | ttnn `grad_y` | torch |
|---|---|---|---|
| 2 | 1 | +inf | +inf |
| -2 | 1 | **+inf** | -inf |
| 0 | 1 | **+inf** | NaN |
| -2 | -1 | **-inf** | +inf |

`addcdiv_bw`, `tensor2 == 0`, `value = 1`:

| grad | tensor1 | ttnn `grad_t1` | torch | ttnn `grad_t2` | torch |
|---|---|---|---|---|---|
| -1 | 2 | **+inf** | -inf | **-inf** | +inf |
| 1 | -2 | +inf | +inf | **-inf** | +inf |
| 1 | 0 | +inf | +inf | **-inf** | NaN |
| 1 | 2 | +inf | +inf | -inf | -inf |

With `value = -1` every `grad_t1` sign is wrong except `grad = -1, tensor1 = 2`.

## Why the tests do not catch it

The nightly tests and sweeps draw the divisor from `uniform(0.001, 100)` or `uniform(-100, 100)` in bf16, so an exact zero essentially never occurs, and the PCC comparison ignores non-finite lanes.

Distinct from #52376 (range loss from inverting the divisor's square, which also affects `addcdiv_bw`'s `tensor2` gradient but is about magnitude, not sign), #53872 (dead `eq(x, nan)` guards in `xlogy_bw`) and #54824 (gradient selection).

## Reproducer (host only, `pip install torch`)

```python
import torch
INF, NAN = float("inf"), float("nan")

# xlogy_bw, grad wrt y at y == 0 (binary_backward.cpp L295-306)
x = torch.tensor([2., -2., 0., -2.]); y = torch.zeros(4); g = torch.tensor([1., 1., 1., -1.])
ttnn_grad_y = torch.where(y == 0, torch.sign(g) * INF, g * x / y)
xr, yr = x.clone().requires_grad_(True), y.clone().requires_grad_(True)
torch.xlogy(xr, yr).backward(g)
print(ttnn_grad_y.tolist())   # [inf, inf, inf, -inf]
print(yr.grad.tolist())       # [inf, -inf, nan, inf]

# addcdiv_bw at tensor2 == 0, value = 1 (ternary_backward.cpp L57-61, L72-76)
g = torch.tensor([-1., 1., 1., 1.]); t1 = torch.tensor([2., -2., 0., 2.], requires_grad=True)
t2 = torch.zeros(4, requires_grad=True)
torch.addcdiv(torch.zeros(4), t1, t2, value=1.0).backward(g)
print("ttnn grad_t1: [inf, inf, inf, inf]   torch:", t1.grad.tolist())     # [-inf, inf, inf, inf]
print("ttnn grad_t2: [-inf, -inf, -inf, -inf] torch:", t2.grad.tolist())   # [inf, inf, nan, -inf]
```

## Proposed fix

Mirror `div_bw`:

```cpp
// xlogy_bw, other gradient
grad2 = where(eqz(other),
              where(eqz(input_a), nan, multiply(multiply(sign(grad), sign(input_a)), inf)),
              grad2);

// addcdiv_bw
float sv = value > 0 ? 1.f : (value < 0 ? -1.f : 0.f);
grad_a = where(eqz(tensor2), where(eqz(grad), nan, multiply(sign(grad), sv * inf)), grad_a);
grad_b = where(eqz(tensor2),
               where(logical_or(eqz(grad), eqz(tensor1)), nan,
                     multiply(multiply(sign(grad), sign(tensor1)), -sv * inf)),
               grad_b);
```

Add a nightly row with an exact-zero divisor and mixed signs to `test_backward_xlogy.py` and `test_backward_addcdiv.py`.

## Acceptance criteria (proposed)

- At a zero divisor, `xlogy_bw` and `addcdiv_bw` return infinities with the sign torch autograd produces, and NaN where torch produces NaN, for fp32 and bf16.
- Non-zero divisors unchanged (existing PCC tests still pass).
- Nightly tests extended with zero-divisor, mixed-sign rows.

**Comment by ping-dev-ui** [comment]
Reporter here. Fix with tests is ready on my fork (https://github.com/ping-dev-ui/tt-metal/tree/fix/xlogy-addcdiv-bw-zero-divisor): both ops now mirror div_bw, the infinity at a zero divisor takes sign(grad) * sign(numerator) (and the sign of value for addcdiv_bw), with a NaN arm when the numerator is zero, plus nightly rows with an exact-zero divisor and mixed signs. Holding the PR until assignment per the bounty terms. @jberkowitzTT could you assign this to me when it is reviewed?

**Comment by Jashandeep5911** [comment]
Hi, I’d like to work on this issue. Could you please confirm that it is still unclaimed, that the bounty is $750, how payment is delivered, and whether contributors located in India are eligible? I can submit the documentation correction after confirmation.

**Comment by ping-dev-ui** [comment]
I have read and agree to the Tenstorrent Bounty Program Terms & Conditions (https://docs.tenstorrent.com/bounty_terms.html).

**Comment by ZacLou** [comment]
I'm working on this — will open a PR in the next 24h.

**Comment by jberkowitzTT** [comment]
Hi @ping-dev-ui - please submit all bounty proposals to [bounties@tenstorrent.com](mailto:bounties@tenstorrent.com). If approved, we'll create the issue and assign it to you as a bounty

Thank you for your enthusiasm & understanding
